### DISCUSSION TITLE
Injeção de dependência - Múltiplas implementações

### DISCUSSION THREAD

**Discussion by rafaelevoinfo** [balta-io/forum#307] [Tira Dúvidas] [answered]
No Curso "Dominando a injeção de dependência" na aula " Registrando mais de uma implementação" fala sobre como registrar múltiplas implementações para uma mesma interface, porém fiquei com uma dúvida. Como recupero uma instância ou outra dependendo de alguma regra?
Explicando melhor:
Imagine uma interface IFreteService e duas implementações PacService e TransportadoraService. Dependendo de alguma regra variável eu hora vou precisar usar a implementação PacService, hora vou usar a TransportadoraService. 
Existe alguma forma de aplicar alguma regra no momento de pegar uma instância de IFreteService? ou então durante seu registro?

**Comment by andrebaltieri** [comment, marked as answer]
Bom dia, @rafaelevoinfo , como vai? É uma ótima pergunta 😄 


```csharp
// Supondo que tenha registrado suas implementações para uma mesma interface
services.AddTransient<IInterface, Implementation1>();
services.AddTransient<IInterface, Implementation2>();
services.AddTransient<IInterface, Implementation3>();

// Você pode receber uma lista destes serviços
public Consumer(IEnumerable<IInterface> services)
{
  // Iterar sobre eles
   foreach(var service in services)
   {
      // Verificar seu tipo
      var objType = service.GetType();
   }
}
```

Neste caso, acredito que caberia um extension method para fazer esta extração de forma mais fácil.

**Reply by rafaelevoinfo** [reply]
A questão é que dessa forma o código que fosse utilizar teria que saber qual implementação utilizar, o que não me parece fazer muito sentido já que estamos utilizando abstração. 

É possível após a aplicação ter sido iniciada trocar a implementação? ou seja, remover a atual e adicionar uma nova. Porque caso sim uma ideia seria fazer isso no momento X onde minha aplicação detectasse que a regra mudou e que daquele ponto em diante a implementação TransportadoraService precisa ser utilizada ao invés PacService.

**Reply by andrebaltieri** [reply]
Bom dia, @rafaelevoinfo, como vai?

Você quer instanciar um objeto baseado na sua herança? (Factory/Abstract Factory)
Acredito que sejam dois pontos diferentes a se tratar, um mais relacionado a OOP e outro a DI, o ponto de DI está descrito acima e o de OOP você pode ver no curso de Mult-Tenant!
👉 https://balta.io/cursos/aplicacoes-mult-tenant-entity-framework-core
(Acredito que na aula OOP e múltiplas regras de negócio)
